src/platform_impl/linux/common/xkb_state.rs

```diff
@@ -0,0 +1,675 @@
+                    location,
+                ),
+            })
+    }
+
+    pub fn key_without_modifiers(&mut self) -> (Key<'static>, KeyLocation) {
+        // This will become a pointer to an array which libxkbcommon owns, so we don't need to deallocate it.
+        let mut keysyms = ptr::null();
+        let keysym_count = unsafe {
+            (XKBH.xkb_keymap_key_get_syms_by_level)(
+                self.state.xkb_keymap,
+                self.keycode,
+                0,
+                0,
+                &mut keysyms,
+            )
+        };
+        let keysym = if keysym_count == 1 {
+            unsafe { *keysyms }
+        } else {
+            0
+        };
+        self.keysym_to_key(keysym)
+            .unwrap_or_else(|(key, location)| {
+                (
+                    self.state
+                        .keysym_to_utf8_raw(keysym)
+                        .map(Key::Character)
+                        .unwrap_or(key),
+                    location,
+                )
+            })
+    }
+
+    fn keysym_to_key(
+        &mut self,
+        keysym: u32,
+    ) -> Result<(Key<'static>, KeyLocation), (Key<'static>, KeyLocation)> {
+        let location = super::keymap::keysym_location(keysym);
+        let key = super::keymap::keysym_to_key(keysym);
+        if matches!(key, Key::Unidentified(_)) {
+            Err((key, location))
+        } else {
+            Ok((key, location))
+        }
+    }
+
+    pub fn text(&mut self) -> Option<&'static str> {
+        self.composed_text()
+            .unwrap_or_else(|_| self.state.keysym_to_utf8_raw(self.keysym))
+    }
+
+    pub fn text_with_all_modifiers(&mut self) -> Option<&'static str> {
+        // The current behaviour makes it so composing a character overrides attempts to input a
+        // control character with the `Ctrl` key. We can potentially add a configuration option
+        // if someone specifically wants the oppsite behaviour.
+        self.composed_text()
+            .unwrap_or_else(|_| self.state.get_utf8_raw(self.keycode))
+    }
+
+    fn composed_text(&mut self) -> Result<Option<&'static str>, ()> {
+        if let Some(compose) = &self.compose {
+            match compose {
+                XkbCompose::Accepted(status) => match status {
+                    ffi::xkb_compose_status::XKB_COMPOSE_COMPOSED => {
+                        Ok(self.state.compose_get_utf8_normal())
+                    }
+                    ffi::xkb_compose_status::XKB_COMPOSE_NOTHING => Err(()),
+                    _ => Ok(None),
+                },
+                XkbCompose::Ignored | XkbCompose::Uninitialized => Err(()),
+            }
+        } else {
+            Err(())
+        }
+    }
+}
+
+fn byte_slice_to_cached_string(bytes: &[u8]) -> &'static str {
+    use std::cell::RefCell;
+    use std::collections::HashSet;
+
+    thread_local! {
+        static STRING_CACHE: RefCell<HashSet<&'static str>> = RefCell::new(HashSet::new());
```

Comment:
A potential alternative to this system that would avoid this hackiness would be to use `Arc<str>` or `Rc<str>` instead of `&'static str` here. Would that work with the rest of the API?

Reply:
I think we agreed to use SSO instead of all of that, given that event enum is already large enough to hold SSO.